File: evita/dom/components/syntax_highlighter/scripts/tokens.py
# Copyright (c) 2016 Project Vogue. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.


class Token(object):
    def __init__(self, start, end):
        self._end = end
        self._start = start

    @property
    def is_action(self):
        return False

    @property
    def is_char(self):
        return False

    @property
    def is_label(self):
        return False

    @property
    def is_name(self):
        return False

    @property
    def is_operator(self):
        return False

    @property
    def start(self):
        return self._start

    def is_operator_of(self, char_code):
        assert char_code != 0
        return False


class ActionToken(Token):
    def __init__(self, name, start, end):
        super(ActionToken, self).__init__(start, end)
        self._name = name

    @property
    def is_action(self):
        return True

    @property
    def name(self):
        return self._name

    def __str__(self):
        return 'ActionToken(%s)' % self.name


class CharToken(Token):
    def __init__(self, char_code, start, end):
        super(CharToken, self).__init__(start, end)
        self._char_code = char_code

    @property
    def char_code(self):
        return self._char_code

    @property
    def name(self):
        if self.char_code == ord('\''):
            return "'\\''"
        if self.char_code == ord('\\'):
            return "'\\\\'"
        return "'%s'" % chr(self.char_code)

    @property
    def is_char(self):
        return True

    def __str__(self):
        return "CharToken(%s)" % self.name


class LabelToken(Token):
    def __init__(self, name, start, end):
        super(LabelToken, self).__init__(start, end)
        self._name = name

    @property
    def is_label(self):
        return True

    @property
    def name(self):
        return self._name

    def __str__(self):
        return 'LabelToken(%s)' % self.name


class NameToken(Token):
    def __init__(self, name, start, end):
        super(NameToken, self).__init__(start, end)
        self._name = name

    @property
    def is_name(self):
        return True

    @property
    def name(self):
        return self._name

    def __str__(self):
        return 'NameToken(%s)' % self.name


class OperatorToken(Token):
    def __init__(self, char_code, start, end):
        super(OperatorToken, self).__init__(start, end)
        self._char_code = char_code

    @property
    def char_code(self):
        return self._char_code

    @property
    def is_operator(self):
        return True

    def is_operator_of(self, char_code):
        return self.char_code == char_code

    def __str__(self):
        return 'OperatorToken(%s)' % chr(self.char_code)

File: joana/public/ast/tokens.cc
// Copyright (c) 2016 Project Vogue. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "joana/public/ast/tokens.h"

#include "joana/public/ast/error_codes.h"

namespace joana {
namespace ast {

//
// Comment
//
Comment::Comment(const SourceCodeRange& range) : Token(range) {}

Comment::~Comment() = default;

//
// Invalid
//
Invalid::Invalid(const SourceCodeRange& range, int error_code)
    : Token(range), error_code_(error_code) {}

Invalid::~Invalid() = default;

// Implements |Node| members
void Invalid::PrintMoreTo(std::ostream* ostream) const {
  static const char* kLiteralErrorTexts[] = {
#define V(token, reason) "LEXER_" #token "_" #reason,
      FOR_EACH_LEXER_ERROR_CODE(V)
#undef V
  };

  static const char* kParserErrorText[] = {
#define V(token, reason) "PASER_" #token "_" #reason,
      FOR_EACH_LEXER_ERROR_CODE(V)
#undef V
  };

  *ostream << ", ";
  const auto lexer_it =
      std::begin(kLiteralErrorTexts) + error_code_ - kLexerErrorCodeBase - 1;
  if (lexer_it >= std::begin(kLiteralErrorTexts) &&
      lexer_it < std::end(kLiteralErrorTexts)) {
    *ostream << *lexer_it;
    return;
  }
  const auto parser_it =
      std::begin(kParserErrorText) + error_code_ - kParserErrorCodeBase - 1;
  if (parser_it >= std::begin(kParserErrorText) &&
      parser_it < std::end(kParserErrorText)) {
    *ostream << *parser_it;
    return;
  }
  *ostream << error_code_;
}

//
// Name
//
Name::Name(const SourceCodeRange& range, int number)
    : Token(range), number_(number) {}

Name::~Name() = default;

bool Name::IsKeyword() const {
  return number_ > static_cast<int>(NameId::StartOfKeyword) &&
         number_ < static_cast<int>(NameId::EndOfKeyword);
}

// Implements |Node| members
void Name::PrintMoreTo(std::ostream* ostream) const {
  *ostream << ", " << number_;
}

//
// Punctuator
//
Punctuator::Punctuator(const SourceCodeRange& range, PunctuatorKind kind)
    : Token(range), kind_(kind) {}

Punctuator::~Punctuator() = default;

//
// Token
//
Token::Token(const SourceCodeRange& range) : Node(range) {}
Token::~Token() = default;

}  // namespace ast
}  // namespace joana
